sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuDataProducer.scala

```diff
@@ -0,0 +1,181 @@
+    new WrappedGpuDataProducer(wrapped, modifyAndClose)
+}
+
+class EmptyGpuDataProducer[T] extends GpuDataProducer[T] {
+  override def hasNext: Boolean = false
+
+  override def next: T = throw new NoSuchElementException()
+
+  override def close(): Unit = {}
+}
+
+class SingleGpuDataProducer[T](data: T) extends GpuDataProducer[T] {
+  private var isRead = false
+
+  override def hasNext: Boolean = !isRead
+
+  override def next: T = {
+    isRead = true
+    data
+  }
+
+  override def close(): Unit = {
+    if (!isRead) {
+      data match {
+        case a: AutoCloseable => a.close()
+        case _ =>
+      }
+    }
+  }
+}
+
+class WrappedGpuDataProducer[T, U](
+    wrapped: GpuDataProducer[U],
+    modifyAndClose: U => T) extends GpuDataProducer[T] {
+  override def hasNext: Boolean = wrapped.hasNext
+
+  override def next: T = {
+    modifyAndClose(wrapped.next)
+  }
+
+  override def close(): Unit = {
+    wrapped.close()
+  }
+}
+
+object EmptyTableReader extends EmptyGpuDataProducer[Table]
+
+/**
+ * Provides a transition between a GpuDataProducer[Table] and an Iterator[ColumnarBatch]. Because
+ * of the disconnect in semantics between a GpuDataProducer and generally how we use
+ * an Iterator[ColumnarBatch] pointing to GPU data this will drain the producer, converting the
+ * data to columnar batches, and make them all spillable so the GpuSemaphore can be released
+ * in between each call to next. There is one special case, if there is only one table from
+ * the producer it will not be made spillable on the assumption that the semaphore is already
+ * held and will not be released before the first table is consumed. This is also fitting with
+ * the semantics of how we use an Iterator[ColumnarBatch] pointing to GPU data.
+ */
+class CachingGpuBatchIterator(
+    producer: GpuDataProducer[Table],
+    dataTypes: Array[DataType],
+    spillCallback: SpillCallback) extends GpuColumnarBatchIterator(true) with Arm {
+
+  private[this] var notSpillableBatch: Option[ColumnarBatch] = None
+  private[this] val pending: mutable.Queue[SpillableColumnarBatch] = mutable.Queue.empty
+
+  private[this] def makeSpillableAndClose(table: Table): SpillableColumnarBatch = {
+    withResource(table) { _ =>
+      SpillableColumnarBatch(GpuColumnVector.from(table, dataTypes),
+        SpillPriorities.ACTIVE_ON_DECK_PRIORITY,
+        spillCallback)
+    }
+  }
+
+  { // Constructor...
+    withResource(producer) { _ =>
+      if (producer.hasNext) {
+        // Special case for the first one.
+        closeOnExcept(producer.next) { firstTable =>
+          if (!producer.hasNext) {
+            notSpillableBatch = Some(GpuColumnVector.from(firstTable, dataTypes))
+            firstTable.close()
+          } else {
+            pending += makeSpillableAndClose(firstTable)
+            producer.foreach { t =>
+              pending += makeSpillableAndClose(t)
+            }
+          }
+        }
+      }
+    }
+  }
```

Comment:
This is the code I was referring to before that I feel should be made into an apply method so this class doesn't need to deal with the two distinct cases of a singleton unspillable batch vs. a string of cached batches.  It would look something like this:
```scala
object CachingGpuBatchIterator {
  private def makeSpillableAndClose(
      table: Table,
      spillCallback: SpillCallback): SpillableColumnarBatch = {
    withResource(table) { _ =>
      SpillableColumnarBatch(GpuColumnVector.from(table, dataTypes),
        SpillPriorities.ACTIVE_ON_DECK_PRIORITY,
        spillCallback)
    }
  }

  def apply(
      producer: GpuDataProducer[Table],
      dataTypes: Array[DataType],
      spillCallback: SpillCallback): GpuColumnarBatchIterator = {
    withResource(producer) { _ =>
      if (producer.hasNext) {
        // Special case for the first one.
        val firstTable = producer.next
        val hasMore = closeOnExcept(firstTable)(producer.hasNext)
        if (!hasMore) {
          val batch = withResource(firstTable) { _ =>
            GpuColumnVector.from(firstTable, dataTypes)
          }
          new SingleGpuColumnarBatchIterator(batch)
        } else {
          val pending = closeOnExcept(mutable.Queue.empty[SpillableColumnarBatch]) { pending =>
            pending += makeSpillableAndClose(firstTable, spillCallback)
            producer.foreach { t =>
              pending += makeSpillableAndClose(t)
            }
          }
          new CachingGpuBatchIterator(pending, 
        }
      } else {
        EmptyGpuColumnarBatchIterator
      }
    }
  }
}

class CachingGpuBatchIterator private (
    pending: mutable.Queue[SpillableColumnarBatch]) {
  override def hasNext: Boolean = pending.nonEmpty

  override def next(): ColumnarBatch = {
    withResource(pending.dequeue()) { spillable =>
      spillable.getColumnarBatch()
    }
  }

  override def doClose(): Unit = {
    pending.safeClose()
  }
}
```